crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -641,6 +734,56 @@ impl AuthorityPerEpochStore {
             .get(&FINAL_EPOCH_CHECKPOINT_INDEX)?)
     }
 
+    /// Returns transaction digests from consensus_message_order table in the "checkpoint range".
+    ///
+    /// Checkpoint range is defined from the last seen checkpoint(excluded) to the provided
+    /// to_height (included)
+    pub fn last_checkpoint(
+        &self,
+        to_height_included: u64,
+    ) -> SuiResult<Option<(u64, Vec<TransactionDigest>)>> {
+        let (index, from_height_excluded) = self.get_last_checkpoint_boundary();
+
+        if let Some(from_height_excluded) = from_height_excluded {
+            if from_height_excluded >= to_height_included {
+                // Due to crash recovery we might enter this function twice for same boundary
+                debug!("Not returning last checkpoint - already processed");
+                return Ok(None);
+            }
+        }
+
+        let roots =
+            self.get_transactions_in_checkpoint_range(from_height_excluded, to_height_included)?;
+
+        debug!(
+            "Selected {} roots between narwhal commit rounds {:?} and {}",
+            roots.len(),
+            from_height_excluded,
+            to_height_included
+        );
+
+        Ok(Some((index, roots)))
+    }
+    pub fn record_checkpoint_boundary(&self, commit_round: u64) -> SuiResult {
+        let (index, height) = self.get_last_checkpoint_boundary();
+
+        if let Some(height) = height {
+            if height >= commit_round {
+                // Due to crash recovery we might see same boundary twice
+                debug!("Not recording checkpoint boundary - already updated");
+                return Ok(());
+            }
+        }
+
+        let index = index + 1;
+        debug!(
+            "Recording checkpoint boundary {} at {}",
+            index, commit_round
+        );
+        self.insert_checkpoint_boundary(index, commit_round)?;
+        Ok(())
+    }
```

Comment:
i'm assuming all of this was moved without significant changes?

Reply:
Right. All moved from the store into the per epoch store